File: sdks/sandbox/python/src/opensandbox/config/connection.py
#
# Copyright 2025 Alibaba Group Holding Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""
Connection configuration for OpenSandbox operations.
"""

import os
from datetime import timedelta

import httpx  # type: ignore[reportMissingImports]
from pydantic import (  # type: ignore[reportMissingImports]
    BaseModel,
    ConfigDict,
    Field,
    PrivateAttr,
    field_validator,
)

from opensandbox.transport import RetryAsyncTransport, RetryPolicy


class ConnectionConfig(BaseModel):
    """
    Sandbox operations connection configuration.

    Transport lifecycle:
    - If `transport` is NOT provided, the SDK creates a default `httpx.AsyncHTTPTransport`
      per Sandbox/Manager instance. In this case, `Sandbox.close()` / `SandboxManager.close()`
      will close the transport.
    - If `transport` IS provided by the user, the SDK will NOT close it; the user owns it.

    Note:
    - Async transports are generally expected to be used within a single asyncio event loop.
      If your test runner creates multiple event loops (common in pytest-asyncio defaults),
      avoid sharing a single `ConnectionConfig(transport=...)` instance across loops.
    """

    model_config = ConfigDict(arbitrary_types_allowed=True)

    _owns_transport: bool = PrivateAttr(default=True)

    api_key: str | None = Field(
        default=None, description="API key for authentication with sandbox service"
    )
    domain: str | None = Field(
        default=None, description="Base domain for the sandbox management API"
    )
    protocol: str = Field(default="http", description="Protocol to use (http/https)")
    request_timeout: timedelta = Field(
        default=timedelta(seconds=30),
        description="Timeout for HTTP requests to the management API",
    )
    debug: bool = Field(
        default=False, description="Enable debug logging for HTTP requests"
    )
    user_agent: str = Field(
        default="OpenSandbox-Python-SDK/0.1.17.dev0", description="User agent string"
    )
    headers: dict[str, str] = Field(
        default_factory=dict, description="User defined headers"
    )
    transport: httpx.AsyncBaseTransport | None = Field(
        default=None,
        description=(
            "Shared httpx transport instance used by all HTTP clients within a "
            "Sandbox/Manager instance. When unset the SDK builds an "
            "AsyncHTTPTransport wrapped by RetryAsyncTransport honoring "
            "`retry_policy`. Pass a custom transport to override the whole "
            "stack (retry wrapping is then the caller's job)."
        ),
    )
    retry_policy: RetryPolicy = Field(
        default_factory=RetryPolicy,
        description=(
            "Retry policy applied to non-streaming requests going through "
            "the shared transport. Pass RetryPolicy.disabled() for fast-fail."
        ),
    )
    use_server_proxy: bool = Field(
        default=False,
        description=(
            "Using sandbox server as proxy for process execd requests"
            "It's useful when client sdk can't access the created sandbox directly"
        ),
    )
    endpoint_cache_ttl: timedelta = Field(
        default=timedelta(seconds=600),
        description="TTL for cached endpoint entries.",
    )
    endpoint_cache_size: int = Field(
        default=1024,
        description="Maximum number of cached endpoint entries. 0 means default (1024).",
    )
    endpoint_cache_disabled: bool = Field(
        default=False,
        description="Disable endpoint caching entirely.",
    )
    disable_metrics: bool = Field(
        default=False,
        description=(
            "Disable SDK telemetry (sandbox.create latency reports). "
            "Also honored via OPENSANDBOX_DISABLE_METRICS=1."
        ),
    )
    enable_tracing: bool = Field(
        default=False,
        description="Enable OpenTelemetry tracing for SDK operations.",
    )

    # Environment variable names
    _ENV_API_KEY = "OPEN_SANDBOX_API_KEY"
    _ENV_DOMAIN = "OPEN_SANDBOX_DOMAIN"
    _DEFAULT_DOMAIN = "localhost:8080"
    _API_VERSION = "v1"

    def model_post_init(self, __context: object) -> None:
        # If the user explicitly provided `transport`, the SDK must not close it.
        self._owns_transport = "transport" not in self.model_fields_set
        # Best-effort: attach the SDK host's own IP so the server can see the
        # client's self-reported address. Never overrides a user-supplied value
        # and is skipped silently when the IP cannot be determined.
        from opensandbox.config import client_ip

        client_ip.apply_client_ip(self.headers)

    def with_transport_if_missing(
        self,
        *,
        max_connections: int = 100,
        max_keepalive_connections: int = 20,
        keepalive_expiry: float = 30.0,
    ) -> "ConnectionConfig":
        """
        Ensure a transport exists for this SDK resource.

        When `transport` is missing, return a copy whose `transport` is
        a retry-wrapped AsyncHTTPTransport (unless the policy has no
        wrapper-only knobs, in which case the raw transport is used).
        When `transport` is present, return self unchanged; the caller
        owns retry wrapping.
        """
        if self.transport is not None:
            return self
        inner = httpx.AsyncHTTPTransport(
            limits=httpx.Limits(
                max_connections=max_connections,
                max_keepalive_connections=max_keepalive_connections,
                keepalive_expiry=keepalive_expiry,
            ),
        )
        wrapped: httpx.AsyncBaseTransport
        if self.retry_policy.wraps_transport():
            wrapped = RetryAsyncTransport(inner, self.retry_policy, owns_inner=True)
        else:
            wrapped = inner
        config = self.model_copy(update={"transport": wrapped})
        config._owns_transport = True
        return config

    async def close_transport_if_owned(self) -> None:
        """Close the transport only if it was created by default_factory."""
        if self.transport is None or not self._owns_transport:
            return
        try:
            await self.transport.aclose()
        except Exception:
            # Avoid raising during cleanup paths.
            pass

    @field_validator("protocol")
    @classmethod
    def protocol_must_be_valid(cls, v: str) -> str:
        v = v.lower()
        if v not in ("http", "https"):
            raise ValueError("Protocol must be 'http' or 'https'")
        return v

    @field_validator("request_timeout")
    @classmethod
    def timeout_must_be_positive(cls, v: timedelta) -> timedelta:
        if v.total_seconds() <= 0:
            raise ValueError(f"Request timeout must be positive, got: {v}")
        return v

    def get_api_key(self) -> str:
        """
        Get API key from config or environment variable.
        Returns:
            API key string (may be empty if not configured)
        Note: An empty API key may cause authentication failures.
        Consider checking if the key is set before making API calls.
        """
        return self.api_key or os.getenv(self._ENV_API_KEY, "")

    def get_domain(self) -> str:
        """Get domain from config or environment variable."""
        return self.domain or os.getenv(self._ENV_DOMAIN, self._DEFAULT_DOMAIN)

    def get_base_url(self) -> str:
        """Get the full base URL for API requests."""
        domain = self.get_domain()
        # Allow domain to override protocol if it explicitly starts with a scheme
        if domain.startswith("http://") or domain.startswith("https://"):
            return f"{domain}/{self._API_VERSION}"
        return f"{self.protocol}://{domain}/{self._API_VERSION}"

File: sdks/sandbox/javascript/src/config/connection.ts
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {DEFAULT_USER_AGENT} from "../core/constants.js";
import {ensureClientIpReady, withClientIp} from "./clientIp.js";

export type ConnectionProtocol = "http" | "https";

/**
 * Options for {@link ConnectionConfig}.
 *
 * Most users only need `domain`, `protocol`, and `apiKey`.
 */
export interface ConnectionConfigOptions {
  /**
   * API server domain (host[:port]) without scheme.
   * Examples:
   * - "localhost:8080"
   * - "api.opensandbox.io"
   *
   * You may also pass a full URL (e.g. "http://localhost:8080" or "https://api.example.com").
   * If the URL includes a path, it will be preserved and `/v1` will be appended automatically.
   */
  domain?: string;
  protocol?: ConnectionProtocol;
  apiKey?: string;
  headers?: Record<string, string>;

  /**
   * Request timeout applied to all SDK HTTP calls (best-effort; wraps fetch).
   * Defaults to 30 seconds.
   */
  requestTimeoutSeconds?: number;
  /**
   * Enable basic debug logging for HTTP requests (best-effort).
   */
  debug?: boolean;
  /**
   * Use sandbox server as proxy for process execd requests.
   * Useful when the client SDK cannot access the created sandbox directly.
   */
  useServerProxy?: boolean;
  /**
   * TTL in milliseconds for cached endpoint entries. Default: 600000 (10 minutes).
   */
  endpointCacheTtlMs?: number;
  /**
   * Maximum number of cached endpoint entries. Default: 1024.
   */
  endpointCacheSize?: number;
  /**
   * Disable endpoint caching entirely.
   */
  endpointCacheDisabled?: boolean;
  /**
   * Disable SDK telemetry (sandbox.create latency reports).
   * Also honored via `OPENSANDBOX_DISABLE_METRICS=1`.
   */
  disableMetrics?: boolean;
  /**
   * Enable OpenTelemetry tracing for client-side pool warmup.
   * Off by default.
   */
  enableTracing?: boolean;
}

function isNodeRuntime(): boolean {
  const p = (globalThis as any)?.process;
  return !!p?.versions?.node;
}

function redactHeaders(
  headers: Record<string, string>
): Record<string, string> {
  const out: Record<string, string> = { ...headers };
  for (const k of Object.keys(out)) {
    if (k.toLowerCase() === "open-sandbox-api-key") out[k] = "***";
  }
  return out;
}

function readEnv(name: string): string | undefined {
  const env = (globalThis as any)?.process?.env;
  const v = env?.[name];
  return typeof v === "string" && v.length ? v : undefined;
}

function stripTrailingSlashes(s: string): string {
  let end = s.length;
  while (end > 0 && s.charCodeAt(end - 1) === 47) {
    end -= 1;
  }
  return end === s.length ? s : s.slice(0, end);
}

function stripV1Suffix(s: string): string {
  const trimmed = stripTrailingSlashes(s);
  return trimmed.endsWith("/v1") ? trimmed.slice(0, -3) : trimmed;
}

const DEFAULT_KEEPALIVE_TIMEOUT_MS = 30_000;

function normalizeDomainBase(input: string): {
  protocol?: ConnectionProtocol;
  domainBase: string;
} {
  // Accept a full URL and preserve its path prefix (if any).
  if (input.startsWith("http://") || input.startsWith("https://")) {
    const u = new URL(input);
    const proto = u.protocol === "https:" ? "https" : "http";
    // Keep origin + pathname, drop query/hash.
    const base = `${u.origin}${u.pathname}`;
    return { protocol: proto, domainBase: stripV1Suffix(base) };
  }

  // No scheme: treat as "host[:port]" or "host[:port]/prefix" and normalize trailing "/v1" or "/".
  return { domainBase: stripV1Suffix(input) };
}

function createNodeFetch(): {
  fetch: typeof fetch;
  close: () => Promise<void>;
} {
  if (!isNodeRuntime()) {
    return {
      fetch,
      close: async () => {
        // Browser fetch has no pooled dispatcher to close.
      },
    };
  }

  const baseFetch = fetch;
  let dispatcher: unknown | undefined;
  let dispatcherPromise: Promise<unknown> | null = null;

  const nodeFetch: typeof fetch = async (input: RequestInfo | URL, init?: RequestInit) => {
    dispatcherPromise ??= (async () => {
      try {
        const mod = await import("undici");
        const Agent = (mod as { Agent?: new (...args: any[]) => unknown }).Agent;
        if (!Agent) {
          return undefined;
        }
        dispatcher = new Agent({
          keepAliveTimeout: DEFAULT_KEEPALIVE_TIMEOUT_MS,
          keepAliveMaxTimeout: DEFAULT_KEEPALIVE_TIMEOUT_MS,
        });
        return dispatcher;
      } catch {
        return undefined;
      }
    })();

    if (dispatcherPromise) {
      await dispatcherPromise;
    }

    if (dispatcher) {
      const mergedInit = { ...(init ?? {}), dispatcher } as RequestInit & {
        dispatcher?: unknown;
      };
      return baseFetch(input, mergedInit as RequestInit);
    }

    return baseFetch(input, init);
  };

  return {
    fetch: nodeFetch,
    close: async () => {
      if (dispatcherPromise) {
        await dispatcherPromise.catch(() => undefined);
      }
      if (
        dispatcher &&
        typeof dispatcher === "object" &&
        typeof (dispatcher as any).close === "function"
      ) {
        try {
          await (dispatcher as any).close();
        } catch {
          // swallow close errors
        }
      }
    },
  };
}

function createTimedFetch(opts: {
  baseFetch: typeof fetch;
  timeoutSeconds: number;
  debug: boolean;
  defaultHeaders?: Record<string, string>;
  label: string;
}): typeof fetch {
  const baseFetch = opts.baseFetch;
  const timeoutSeconds = opts.timeoutSeconds;
  const debug = opts.debug;
  const defaultHeaders = opts.defaultHeaders ?? {};
  const label = opts.label;

  return async (input: RequestInfo | URL, init?: RequestInit) => {
    const method = init?.method ?? "GET";
    const url =
      typeof input === "string"
        ? input
        : (input as any)?.toString?.() ?? String(input);

    const ac = new AbortController();
    const requestSignal =
      init?.signal ??
      (typeof Request !== "undefined" && input instanceof Request
        ? input.signal
        : undefined);
    const timeoutMs = Math.floor(timeoutSeconds * 1000);
    const t =
      Number.isFinite(timeoutMs) && timeoutMs > 0
        ? setTimeout(
            () =>
              ac.abort(
                new Error(
                  `[${label}] Request timed out (timeoutSeconds=${timeoutSeconds})`
                )
              ),
            timeoutMs
          )
        : undefined;

    const onAbort = () =>
      ac.abort((requestSignal as any)?.reason ?? new Error("Aborted"));
    if (requestSignal) {
      if (requestSignal.aborted) onAbort();
      else
        requestSignal.addEventListener("abort", onAbort, { once: true } as any);
    }

    // Best-effort: attach the SDK host's own IP so the server can see the
    // client's self-reported address. Applied per request (never overriding a
    // caller-supplied value or dropping existing headers) and skipped silently
    // when the IP is unavailable. Await the one-time detection so even the very
    // first request carries the header (bounded by the probe timeout).
    await ensureClientIpReady();
    const withIp = withClientIp(input, init);
    const reqInput = withIp.input;
    const mergedInit: RequestInit = {
      ...(withIp.init ?? {}),
      signal: ac.signal,
    };

    if (debug) {
      // Log the headers actually being sent: prefer the merged headers produced
      // by withClientIp (which include the client-IP header), falling back to
      // the request input's headers when it is a Request object.
      const outgoing = new Headers(
        withIp.init?.headers ??
          (typeof Request !== "undefined" && reqInput instanceof Request
            ? reqInput.headers
            : undefined)
      );
      const mergedHeaders: Record<string, string> = { ...defaultHeaders };
      outgoing.forEach((value, key) => {
        mergedHeaders[key] = value;
      });
      // eslint-disable-next-line no-console
      console.log(
        `[opensandbox:${label}] ->`,
        method,
        url,
        redactHeaders(mergedHeaders)
      );
    }

    try {
      const res = await baseFetch(reqInput, mergedInit);
      if (debug) {
        // eslint-disable-next-line no-console
        console.log(`[opensandbox:${label}] <-`, method, url, res.status);
      }
      return res;
    } finally {
      if (t) clearTimeout(t);
      if (requestSignal)
        requestSignal.removeEventListener("abort", onAbort as any);
    }
  };
}

export class ConnectionConfig {
  readonly protocol: ConnectionProtocol;
  readonly domain: string;
  readonly apiKey?: string;
  readonly headers: Record<string, string>;
  private _fetch: typeof fetch | null;
  private _sseFetch: typeof fetch | null;
  readonly requestTimeoutSeconds: number;
  readonly debug: boolean;
  readonly userAgent: string = DEFAULT_USER_AGENT;
  /**
   * Use sandbox server as proxy for endpoint requests (default false).
   */
  readonly useServerProxy: boolean;
  readonly endpointCacheTtlMs: number;
  readonly endpointCacheSize: number;
  readonly endpointCacheDisabled: boolean;
  readonly disableMetrics: boolean;
  readonly enableTracing: boolean;
  private _closeTransport: () => Promise<void>;
  private _closePromise: Promise<void> | null = null;
  private _transportInitialized = false;

  /**
   * Create a connection configuration.
   *
   * Environment variables (optional):
   * - `OPEN_SANDBOX_DOMAIN` (default: `localhost:8080`)
   * - `OPEN_SANDBOX_API_KEY`
   * - `OPENSANDBOX_DISABLE_METRICS=1` to opt out of create-latency telemetry
   */
  constructor(opts: ConnectionConfigOptions = {}) {
    const envDomain = readEnv("OPEN_SANDBOX_DOMAIN");
    const envApiKey = readEnv("OPEN_SANDBOX_API_KEY");

    const rawDomain = opts.domain ?? envDomain ?? "localhost:8080";
    const normalized = normalizeDomainBase(rawDomain);

    // If the domain includes a scheme, it overrides `protocol`.
    this.protocol = normalized.protocol ?? opts.protocol ?? "http";
    this.domain = normalized.domainBase;
    this.apiKey = opts.apiKey ?? envApiKey;
    this.requestTimeoutSeconds =
      typeof opts.requestTimeoutSeconds === "number"
        ? opts.requestTimeoutSeconds
        : 30;
    this.debug = !!opts.debug;
    this.useServerProxy = !!opts.useServerProxy;
    this.endpointCacheTtlMs = opts.endpointCacheTtlMs ?? 600_000;
    this.endpointCacheSize = opts.endpointCacheSize ?? 1024;
    this.endpointCacheDisabled = !!opts.endpointCacheDisabled;
    this.disableMetrics = !!opts.disableMetrics;
    this.enableTracing = !!opts.enableTracing;

    const headers: Record<string, string> = { ...(opts.headers ?? {}) };
    // Attach API key via header unless the user already provided one.
    if (this.apiKey && !headers["OPEN-SANDBOX-API-KEY"]) {
      headers["OPEN-SANDBOX-API-KEY"] = this.apiKey;
    }
    // Best-effort user-agent (Node only).
    if (
      isNodeRuntime() &&
      this.userAgent &&
      !headers["user-agent"] &&
      !headers["User-Agent"]
    ) {
      headers["user-agent"] = this.userAgent;
    }
    this.headers = headers;
    this._fetch = null;
    this._sseFetch = null;
    this._closeTransport = async () => {
      // Init with empty close call
    };
    this._transportInitialized = false;
  }

  get fetch(): typeof fetch {
    return this._fetch ?? fetch;
  }

  get sseFetch(): typeof fetch {
    return this._sseFetch ?? fetch;
  }

  getBaseUrl(): string {
    // If `domain` already contains a scheme, treat it as a full base URL prefix.
    if (
      this.domain.startsWith("http://") ||
      this.domain.startsWith("https://")
    ) {
      return `${stripV1Suffix(this.domain)}/v1`;
    }
    return `${this.protocol}://${stripV1Suffix(this.domain)}/v1`;
  }

  private initializeTransport(): void {
    if (this._transportInitialized) return;

    const { fetch: baseFetch, close } = createNodeFetch();
    this._fetch = createTimedFetch({
      baseFetch,
      timeoutSeconds: this.requestTimeoutSeconds,
      debug: this.debug,
      defaultHeaders: this.headers,
      label: "http",
    });
    this._sseFetch = createTimedFetch({
      baseFetch,
      timeoutSeconds: 0,
      debug: this.debug,
      defaultHeaders: this.headers,
      label: "sse",
    });
    this._closeTransport = close;
    this._transportInitialized = true;
  }
  /**
   * Ensure this configuration has transport helpers (fetch/SSE) allocated.
   *
   * On Node.js this creates a dedicated `undici` dispatcher; on browsers it
   * simply reuses the global fetch. Returns either `this` or a cloned config
   * with the transport initialized.
   */
  withTransportIfMissing(): ConnectionConfig {
    if (this._transportInitialized) {
      return this;
    }

    const clone = new ConnectionConfig({
      domain: this.domain,
      protocol: this.protocol,
      apiKey: this.apiKey,
      headers: { ...this.headers },
      requestTimeoutSeconds: this.requestTimeoutSeconds,
      debug: this.debug,
      useServerProxy: this.useServerProxy,
      endpointCacheTtlMs: this.endpointCacheTtlMs,
      endpointCacheSize: this.endpointCacheSize,
      endpointCacheDisabled: this.endpointCacheDisabled,
      disableMetrics: this.disableMetrics,
      enableTracing: this.enableTracing,
    });
    clone.initializeTransport();
    return clone;
  }

  /**
   * Close the Node.js agent owned by this configuration.
   */
  async closeTransport(): Promise<void> {
    if (!this._transportInitialized) return;
    this._closePromise ??= this._closeTransport();
    await this._closePromise;
  }
}
